legate/core/shape.py

```diff
@@ -55,57 +65,71 @@ def extents(self):
             self._extents = tuple(hi[idx] + 1 for idx in range(hi.dim))
         return self._extents
 
-    def __str__(self):
+    def __str__(self) -> str:
         if self._extents is not None:
             return f"Shape({self._extents})"
         else:
             return f"Shape({self._ispace})"
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return str(self)
 
-    def __getitem__(self, idx):
+    @overload
+    def __getitem__(self, idx: int) -> int:
+        ...
+
+    @overload
+    def __getitem__(self, idx: slice) -> Shape:
+        ...
+
+    def __getitem__(self, idx: Union[int, slice]) -> Union[Shape, int]:
         if isinstance(idx, slice):
             return Shape(self.extents[idx])
         else:
             return self.extents[idx]
 
-    def __len__(self):
+    def __len__(self) -> int:
         return len(self.extents)
 
+    def __iter__(self) -> Iterator[int]:
+        return iter(self.extents)
+
+    def __contains__(self, value: object) -> bool:
+        return value in self.extents
```

Comment:
Some places expect `Sized[Shape]` and other expect `Iterable[Shape]`. The simplest protocol with both is `Collection[Shape]` but that also requires `__contains__`.